src/napari_spatialdata/utils/_utils.py

```diff
@@ -279,3 +281,29 @@ def _adjust_channels_order(element: SpatialImage | MultiscaleSpatialImage) -> tu
         new_raster = list_of_xdata
 
     return new_raster, rgb
+
+
+def _get_sdata_key(sdata: EventedList, elements: dict[str, dict[str, str | int]], key: str) -> tuple[SpatialData, bool]:
```

Comment:
general question, if the user renames some layers or duplicates some layers, this function will still operate correctly right?

Reply:
No I think it would not, see one of my other comments (the one on `rfind()`)

Reply:
This would still work correctly as this function is only called when adding a layer when loading spatialdata objects by calling Interactive.
The layers are not loaded yet at the stage when this function is called. A user can change the layer name, but as far as I understand we wouldn't necessarily save
that back to SpatialData yet, so the element name is still valid. Just the user would lose the hint
to which SpatialData object it belongs.

Reply:
In other words this function only deals with the elements in the `ElementWidget` and not with layers.

Reply:
I'm not sure I understood this 100%, but I explain the use case I have in mind. Say that we add a shapes layer (polygons) to napari. If we want to plot annotations on top of it we need to know two things: 
1) the `GeoDataFrame` object (because we need the index and maybe also additional columns)
2) the `table` in the associated `SpatialData` object.
This information is not needed just when creating the layer, but any time the user wants to visualize some numerical/categorical annotations.

Saving the index of the `SpatialData` object and the `name` of the element is fragile upon data renaming and removing/adding new `SpatialData` object. I would therefore add as metadata the `GeoDataFrame` (or other type of element) and the original `SpatialData` object.

This implies that we don't need to save `element_type` (since it can be deduced on the fly, and we don't need `sdata_index`. 

Remaining `original_name`, we can store it, or we can maybe deduce it on the fly by removing the last `_[0-9]+$` from the name. Not sure which of these two approaches is best.

Reply:
So currently the layers do have the `SpatialData` object associated with it in the metadata. Changing the layer name has no effect here. Upon saving the `SpatialData` object in the metadata will be used. This would have 1 path associated with it, so even if saving multiple layers sequentially, it does not matter that the metadata for sdata is not updated yet. Or we need to update corresponding `sdata` metadata after each layer is saved, but I think because of using references to one object this is not the case.

The only reason for using the index of `sdata` is to provide the user a clue to which `SpatialData` object in the `EventedList` a layer belongs. The reason for this being that we don't have an identifier that is easily visible for the user when having multiple layers open that would otherwise have a naming collision. Napari's behaviour for naming collisions could lead to confusion for the user.

In this PR we don't give the option yet for adding or removing a `SpatialData` object using a method in `Interactive`. When we do in the follow up PR I think the index of the layer names can be easily determined should there be a need (only in case there are duplicated element names). This can just be done by looking what the index is of `sdata` in metadata in the `EventedList` and then performing a layer rename.

Reply:
Also, element type is already used on main just to deduce what kind of layer should be created. The functionality here did not change other than that now it is a key with a value in a dictionary while before it was directly a value in `_elements`.

Reply:
this is quite complicated to follow in terms of user behaviour and use cases.

> Saving the index of the SpatialData object and the name of the element is fragile upon data renaming and removing/adding new SpatialData object. 

Why would we want the user to rename a napari layer? I would actually even consider if possible to disallow that at the GUI level (prob difficult). it's also in scope with this being an interactive viewer that at most "add" elements but not rename name (especially not in GUI).

> I would therefore add as metadata the GeoDataFrame (or other type of element) and the original SpatialData object.

This doesn't feel right and has the potential to create a lot of chaos imho

Reply:
To answer your question, I don't think we would like the user to do this, except when users add a new layer for ROI annotation. For example, someone opens their image data to annotate artefacts using a shapes layer that they add. By default this layer gets the name shapes. A user would probably like to rename this, then use the linking functionality that I implemented to link the layer to a particular `SpatialData` object and then save it.
The case that I do catch here is when users add multiple spatialdata objects in which some `SpatialElements` share names. Because the original name in the `SpatialData` object is part of `layer.metadata` this is safe to do so. It doesn't have any other effect as far as I can tell then to provide the user with a clue in case of a name collision, which `SpatialData` object the layer belongs to.

One thing to note is that `sdata` is already in `layer.metadata` so adding the GeoPandas df would be overkill.

In any case I do not think the functionality would be broken the way I implemented it currently. I think we are talking about edge cases where we are not even certain whether users would run into this issue. I would vote in this case for first seeing whether users are actually going to run into this issue.

Reply:
Ok thanks for the info, I'd actually go for your approach.